### PR TITLE
refactor: use view engine also for `NgModuleFactory`s

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: packages/compiler/test/aot/regression_spec.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        export class EmptyComp {}
+
+        @NgModule({declarations: [EmptyComp]})
+        export class MyModule {}
+      `
+       };
+       const rootDir = {'app': appDir};
+       compile([rootDir, angularFiles], {postCompile: expectNoDiagnostics}, {
+         noUnusedLocals: false,
+         noUnusedParameters: false
```
Comment (vicb) [inline packages/compiler/test/aot/regression_spec.ts]: should this be true ?

Comment (tbosch) [inline packages/compiler/test/aot/regression_spec.ts]: Yes, of course! Thanks!

**Review by vicb — APPROVED** [review]

**Review by IgorMinar — APPROVED** [review]
lgtm for the public api change. we don't expect anyone to be extending generated code.

File: packages/compiler/src/provider_analyzer.ts [hunk 2]
```diff
@@ -96,7 +96,11 @@ export class ProviderElementContext {
   }
 
   get transformProviders(): ProviderAst[] {
-    return Array.from(this._transformedProviders.values());
+    // Note: Maps keep their insertion order.
+    const providers = Array.from(this._transformedProviders.values());
+    const lazyProviders = providers.filter(p => !p.eager);
```
Comment (vicb) [inline packages/compiler/src/provider_analyzer.ts]: forEach and single iteration ? walking the providers twice is useless

File: packages/compiler/src/provider_analyzer.ts [hunk 3]
```diff
@@ -316,7 +320,11 @@ export class NgModuleProviderAnalyzer {
       const errorString = this._errors.join('\n');
       throw new Error(`Provider parse errors:\n${errorString}`);
     }
-    return Array.from(this._transformedProviders.values());
+    // Note: Maps keep their insertion order.
+    const providers = Array.from(this._transformedProviders.values());
+    const lazyProviders = providers.filter(p => !p.eager);
+    const eagerProviders = providers.filter(p => p.eager);
```
Comment (vicb) [inline packages/compiler/src/provider_analyzer.ts]: same comment

File: packages/core/test/linker/view_injector_integration_spec.ts [hunk 4]
```diff
@@ -342,6 +342,19 @@ export function main() {
         expect(created).toBe(true);
       });
 
+      it('should allow to inject lazy providers via Injector.get from an eager provider that is declared earlier',
```
Comment (vicb) [inline packages/core/test/linker/view_injector_integration_spec.ts]: allow injecting ?

**Comment by vicb** [comment]
LGTM with minor commentd

File: packages/compiler/src/provider_analyzer.ts (hunk 2 above)
Comment (tbosch) [inline packages/compiler/src/provider_analyzer.ts]: Done

File: packages/compiler/src/provider_analyzer.ts (hunk 3 above)
Comment (tbosch) [inline packages/compiler/src/provider_analyzer.ts]: Done

File: packages/core/test/linker/view_injector_integration_spec.ts (hunk 4 above)
Comment (tbosch) [inline packages/core/test/linker/view_injector_integration_spec.ts]: Done

**Pull request merged**